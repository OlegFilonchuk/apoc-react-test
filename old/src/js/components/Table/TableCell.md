## Example of table with TableCell
It uses [TableHeadRow](#tableheadrow) and [Table](#table)

    (
      <Table>
        <TableHeadRow>
          <th>Title</th>
          <th>Title</th>
          <th>Title</th>
        </TableHeadRow>
        <tbody>
          <tr>
            <TableCell>Content</TableCell>
            <TableCell>Content</TableCell>
            <TableCell>Content</TableCell>
          </tr>
          <tr>
            <TableCell>Content</TableCell>
            <TableCell>Content</TableCell>
            <TableCell>Content</TableCell>
          </tr>
          <tr>
            <TableCell>Content</TableCell>
            <TableCell>Content</TableCell>
            <TableCell>Content</TableCell>
          </tr>
        </tbody>
      </Table>
    )

## Example of table with TableCell with alignment by content type
It uses [TableHeadRow](#tableheadrow), [Table](#table) and [InlineEllipsis](#inlineellipsis)

    (
      <Table>
        <TableHeadRow>
          <TableHeadCell textContent>textContent Title</TableHeadCell>
          <TableHeadCell iconOrGraphicalContent> iconOrGraphicalContent Title</TableHeadCell>
          <TableHeadCell numericContent>numericContent Title</TableHeadCell>
        </TableHeadRow>
        <tbody>
          <tr>
            <TableCell>Simple text</TableCell>
            <TableCell><span className="sc-icon-file" /></TableCell>
            <TableCell>123.43</TableCell>
          </tr>
          <tr>
            <TableCell>Different text</TableCell>
            <TableCell>
              <InlineEllipsis width="150px">
                <span className="sc-icon-check sc-success" /> Completed
              </InlineEllipsis>
            </TableCell>
            <TableCell>-23.23 </TableCell>
          </tr>
          <tr>
            <TableCell>Another one</TableCell>
            <TableCell>
              <InlineEllipsis width="150px">
                <span className="sc-grade-c sc-icon-triangle-warning" /> Needs configurations
              </InlineEllipsis>
            </TableCell>
            <TableCell>$34</TableCell>
          </tr>
        </tbody>
      </Table>
    )
