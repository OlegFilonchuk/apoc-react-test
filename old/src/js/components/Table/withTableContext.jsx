import React from 'react';
import propTypes from 'prop-types';

export const CONTEXT_ID = 'APOC_REACT_TABLE_CONTEXT';

export const tableContextPropTypes = propTypes.shape({
  /**
   * @type {function(uuid: string, index:number, contentTypes: CellContentTypes)}
   */
  registerColumn: propTypes.func,
  /**
   * @type {function(uuid: string)}
   */
  unregisterColumn: propTypes.func,
  /**
   * @type {function(index: ?number):CellContentTypes}
   */
  getColumnContentTypes: propTypes.func
});

/**
 * @type {CellContentTypes}
 */
export const defaultContentTypes = {
  dateContent: false,
  iconOrGraphicalContent: false,
  textContent: false,
  numericContent: false
};

export const defaultCellTableProps = {
  registerColumn: () => null,
  unregisterColumn: () => null,
  getColumnContentTypes: () => defaultContentTypes
};

export function withTableContext(Component) {
  function WithTableContext(props, context) {
    return <Component {...props} tableProps={context[CONTEXT_ID]} />;
  }

  WithTableContext.propTypes = Component.propTypes;

  WithTableContext.contextTypes = {
    [CONTEXT_ID]: tableContextPropTypes
  };

  WithTableContext.displayName = `withTableContext${Component.displayName || Component.name}`;

  WithTableContext.WrappedComponent = Component;

  return WithTableContext;
}

/* eslint-disable no-unused-vars */
/**
 * @typedef {{dateContent: boolean,iconOrGraphicalContent: boolean,textContent: boolean, numericContent: boolean}} CellContentTypes
 */
const CellContentTypes = {};
/* eslint-enable no-unused-vars */
