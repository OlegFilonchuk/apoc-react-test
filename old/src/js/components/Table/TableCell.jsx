import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { withTableContext, tableContextPropTypes, defaultCellTableProps } from './withTableContext';

class TableCell extends React.Component {
  setRef = ref => {
    this.ref = ref;
    this.forceUpdate();
  };

  get childIndex() {
    if (!this.ref) {
      return null;
    }

    return Array.from(this.ref.parentElement.children).indexOf(this.ref);
  }

  /**
   * @return {CellContentTypes}
   */
  get contentTypes() {
    return this.props.tableProps.getColumnContentTypes(this.childIndex);
  }

  render() {
    const { children, className, width, tableProps, ...restOfProps } = this.props;
    const style = width
      ? {
          width,
          maxWidth: width
        }
      : {};

    const contentTypes = this.contentTypes;
    const finalClassName = classNames(className, {
      'sc-left-align': contentTypes.textContent,
      'sc-right-align': contentTypes.numericContent,
      'sc-center-align':
        contentTypes.iconOrGraphicalContent || contentTypes.dateContent || !contentTypes.isContentTypePassed
    });

    return (
      <td {...restOfProps} style={style} className={finalClassName} ref={this.setRef}>
        {children}
      </td>
    );
  }
}

TableCell.propTypes = {
  /**
   * Cell's content
   */
  children: PropTypes.node,

  /**
   * List of custom classes added to component
   */
  className: PropTypes.string,

  /**
   * Cell's width
   */
  width: PropTypes.string,

  tableProps: tableContextPropTypes
};

TableCell.defaultProps = {
  children: null,
  className: '',
  width: '',
  tableProps: defaultCellTableProps
};

const TableCellWithContext = withTableContext(TableCell);

export default TableCellWithContext;
