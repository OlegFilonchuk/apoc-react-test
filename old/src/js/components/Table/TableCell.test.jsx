import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import TableCell from './TableCell';
import { defaultContentTypes } from './withTableContext';

const mountCell = ({ mountMethod = mount, props = {} }) =>
  mountMethod(<TableCell.WrappedComponent {...props}>Hello world</TableCell.WrappedComponent>, {
    attachTo: document.createElement('tr')
  });

describe('<TableCell />', () => {
  describe('should match snapshot ', () => {
    it('with passing `width`', () => {
      expect(toJson(mountCell({ mountMethod: shallow, props: { width: '100px' } }))).toMatchSnapshot();
    });

    it('without passing `width`', () => {
      expect(toJson(mountCell({ mountMethod: shallow }))).toMatchSnapshot();
    });
  });

  it('correctly set `className`', () => {
    const cell = mountCell({ props: { className: 'some-extra-class' } });
    const element = cell.find('td');

    expect(element).toHaveClassName('some-extra-class');
  });

  it('correctly set `width`', () => {
    const cell = mountCell({ props: { width: '100px' } });
    const element = cell.find('td');

    expect(element).toHaveStyle('width', '100px');
    expect(element).toHaveStyle('maxWidth', '100px');
  });

  describe('tableProps.getColumnContentTypes', () => {
    let tableProps;

    beforeEach(() => {
      tableProps = {
        getColumnContentTypes: jest.fn(),
        registerColumn: jest.fn(),
        unregisterColumn: jest.fn()
      };
      tableProps.getColumnContentTypes.mockReturnValue(defaultContentTypes);
    });

    describe('TableCell has no siblings', () => {
      let cell;

      beforeEach(() => {
        cell = mountCell({ props: { tableProps } });
      });
      it('is called with 0 - index of parents children node', () => {
        expect(cell.instance().props.tableProps.getColumnContentTypes).toHaveBeenCalledWith(0);
      });
    });

    describe('TableCell is the first of parents children', () => {
      let cell;

      beforeEach(() => {
        const tr = document.createElement('tr');

        tr.appendChild(document.createElement('td'));
        tr.appendChild(document.createElement('td'));

        cell = mount(<TableCell.WrappedComponent tableProps={tableProps}>Hello world</TableCell.WrappedComponent>, {
          attachTo: tr
        });
      });

      it('is called with 0 - index of parents children node', () => {
        expect(cell.instance().props.tableProps.getColumnContentTypes).toHaveBeenCalledWith(0);
      });
    });

    describe('TableCell becomes third child after initial render', () => {
      let cell;

      beforeEach(() => {
        const tr = document.createElement('tr');

        cell = mount(<TableCell.WrappedComponent tableProps={tableProps}>Hello world</TableCell.WrappedComponent>, {
          attachTo: tr
        });

        const tableCellTd = tr.childNodes[0];

        tr.insertBefore(document.createElement('td'), tableCellTd);
        tr.insertBefore(document.createElement('td'), tableCellTd);
      });

      it('is called with arugment 2 - most current child index', () => {
        cell.setProps({ width: '100px' });
        expect(cell.instance().props.tableProps.getColumnContentTypes).toHaveBeenLastCalledWith(2);
      });
    });
  });
});
