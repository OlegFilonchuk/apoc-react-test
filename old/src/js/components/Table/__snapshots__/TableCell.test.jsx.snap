// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableCell /> should match snapshot  with passing \`width\` 1`] = `
<td
  className="sc-center-align"
  style={
    Object {
      "maxWidth": "100px",
      "width": "100px",
    }
  }
>
  Hello world
</td>
`;

exports[`<TableCell /> should match snapshot  without passing \`width\` 1`] = `
<td
  className="sc-center-align"
  style={Object {}}
>
  Hello world
</td>
`;
